homeassistant/components/telegram_bot/__init__.py

```diff
@@ -551,6 +605,149 @@ def _deprecate_timeout(hass: HomeAssistant, service: ServiceCall) -> None:
     )
 
 
+def _build_targets(
+    hass: HomeAssistant, service: ServiceCall
+) -> list[tuple[TelegramBotConfigEntry, int, str | None]]:
+    """Build list of targets where each target is represented by its corresponding config entry, chat ID and notify entity id (if target is a notify entity)."""
+
+    migrate_chat_ids = _warn_chat_id_migration(hass, service)
+
+    targets: list[tuple[TelegramBotConfigEntry, int, str | None]] = []
+
+    # build target list from notify entities using service data: `entity_id`
+
+    referenced = async_extract_referenced_entity_ids(
+        hass, TargetSelection(service.data)
+    )
+    notify_entity_ids = referenced.referenced | referenced.indirectly_referenced
+
+    # parse entity IDs
+    entity_registry = er.async_get(hass)
+    for notify_entity_id in notify_entity_ids:
+        # get config entry from notify entity
+        registry_entry = entity_registry.async_get(notify_entity_id)
+        if not registry_entry:
+            raise ServiceValidationError(
+                translation_domain=DOMAIN,
+                translation_key="action_failed",
+                translation_placeholders={
+                    "error": f"Notify entity not found: {notify_entity_id}"
+                },
+            )
+        notify_config_entry = hass.config_entries.async_get_known_entry(
+            str(registry_entry.config_entry_id)
+        )
+
+        # get chat id from subentry
+        notify_config_subentry = notify_config_entry.subentries[
+            str(registry_entry.config_subentry_id)
+        ]
+        notify_chat_id: int = notify_config_subentry.data[ATTR_CHAT_ID]
+
+        targets.append((notify_config_entry, notify_chat_id, notify_entity_id))
+
+    has_notify_targets = len(targets) > 0
+
+    # build target list using service data: `config_entry_id` and `chat_id`
+
+    config_entry: TelegramBotConfigEntry | None = None
+    if CONF_CONFIG_ENTRY_ID in service.data:
+        # parse config entry from service data
+        config_entry_id: str = service.data[CONF_CONFIG_ENTRY_ID]
+        config_entry = hass.config_entries.async_get_known_entry(config_entry_id)
+    else:
+        # config entry not provided so we try to determine the default
+        config_entries: list[TelegramBotConfigEntry] = (
+            service.hass.config_entries.async_entries(DOMAIN)
+        )
+        if len(config_entries) == 1:
+            config_entry = config_entries[0]
+
+    # parse chat IDs from service data: `chat_id`
+    if config_entry is not None:
+        chat_ids: list[int] = migrate_chat_ids
+        if ATTR_CHAT_ID in service.data:
+            chat_ids = chat_ids + (
```

Comment:
Have we thought about what the use case is for using both parameters (chat_id and entity_id) at the same time? Is that wanted?